### PR TITLE
Make <md-input> react at the same time as <input> & fix race condition

### PR DESCRIPTION
**Pull request by rayfoss**
See #1835 

Reacts exactly as fast as the normal input, without giving it any super powers unlike an "in element" emit. The failing test were failing before these changes... aka, wasn't me.

### CODE REVIEW

File: src/components/MdField/MdFieldMixin.js [hunk 1]
```diff
@@ -22,9 +22,8 @@ export default {
       },
       set (value) {
         if (value.constructor.toString().match(/function (\w*)/)[1].toLowerCase() !== 'inputevent') {
-          this.$nextTick(() => {
-            this.localValue = value
-          })
+          this.$emit('input', value)
+          this.localValue = value
```
Comment (VdustR) [inline src/components/MdField/MdFieldMixin.js]: Hi, thanks for the post issue and PR.

Do you have any idea in this case without `nextTick`?

```vue
<template>
  <div>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="initial" @input="handleInput"></md-input>
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div>
  </div>
</template>

<script>
  export default {
    name: 'TextFields',
    data() {
      return {
        initial: '',
        inputValue: ''
      }
    },
    methods: {
      setValue() {
        this.initial = 'Set Value'
      },
      handleInput(val) {
        this.inputValue = val
      }
    }
  }
</script>
```

Steps:
  - type: 'abc'
  - click the button

before fix: 
```
result: Set Value
```

after fix:
```
result: 123
```

**Comment by rayfoss** [comment]
I do get different results then input yields, but this PR doesn't seem to matter. 

Without PR:
With PR:
mdinput: abc
input: `{ "isTrusted": true}` then `Set Value`

With PR:
mdinput: abc
input: `{ "isTrusted": true}` then `Set Value`

```
<template>
  <div>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="model" @input="handleInput"></md-input>
      <input v-model="model" @input="handleInput" />
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div>
  </div>
</template>

<script>
export default {
  name: 'TextFields',
  data () {
    return {
      model: '',
      inputValue: ''
    }
  },
  methods: {
    setValue () {
      this.model = 'Set Value'
    },
    handleInput (val) {
      this.inputValue = val
    }
  }
}
</script>
```
https://youtu.be/SW2JmwEDy7M

**Comment by rayfoss** [comment]
I think we're at an overall better position with this patch

**Comment by rayfoss** [comment]
`spellcheck="false"` Another "solution"

**Comment by VdustR** [comment]
I'm sorry about that.

As I mentioned in https://github.com/vuematerial/vue-material/pull/1839#discussion_r199359478, there exists problem with this fix. I'm look for a way to resolve them both.

**Comment by rayfoss** [comment]
https://youtu.be/OYdsr90eBYQ

The input event on the built in <input> field doesn't support that use case well either, supporting it like we do now might be the bug.

```
<template>
  <div>
    <h1>MdInput:</h1>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="initial" @input="handleInput"></md-input>
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div><hr>

    <h1>HTML input:</h1>
    <input v-model="initialHTML" @input="handleInputHTML"></md-input>
    <md-button @click="setValueHTML">click</md-button>
    <div>result: {{inputValueHTML}}</div><hr>
  </div>
</template>

<script>
  export default {
    name: 'TextFields',
    data() {
      return {
        initial: '',
        inputValue: '',
        initialHTML: '',
        inputValueHTML: ''
      }
    },
    methods: {
      setValue() {
        this.initial = 'Set Value'
      },
      handleInput(val) {
        this.inputValue = val
      },
      setValueHTML() {
        this.initialHTML = 'Set Value'
      },
      handleInputHTML(val) {
        this.inputValueHTML = val
      }
    }
  }
</script>
```

**Comment by rayfoss** [comment]
Complete docs/app/App.vue demo: 'Set Value' gets set, the todo is fully added.
```
<template>
  <div>
    <h1>MdInput:</h1>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="initial" @input="handleInput"></md-input>
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div><hr>

    <md-toolbar class="md-primary">
        <h1 class="md-title">vue-material T0-D{{yolo%10}}'s</h1>
    </md-toolbar>

     <form class="create md-layout" @submit.prevent="validateTodo">
      <md-card class="md-layout-item md-size-50 md-small-size-100">
        <md-card-header>
          <div class="md-title">To-Do next: {{form.server}}</div>
        </md-card-header>

        <md-card-content>
          <div class="md-layout md-gutter">
            <div class="md-layout-item md-small-size-100">
              <md-field>
                <label for="todo">md-input</label>
                <md-input id="todo" v-model="form.todo" />
              </md-field>
              <div class="md-field md-theme-default">
                <input id="todo" v-model="form.todo" placeholder="html input"/>
              </div>
            </div>
          </div>
        </md-card-content>

        <md-card-actions>
          <md-button type="submit" class="md-primary">Create Todo</md-button>
        </md-card-actions>
      </md-card>
    </form>
  </div>
</template>

<script>
import '../../dist/vue-material.min.css'
import '../../dist/theme/default.css'

export default {
  name: 'TextFields',
  data() {
    return {
      initial: '',
      inputValue: '',
      form: {
        todo: '',
        server: ''
      }
    }
  },
  methods: {
    setValue() {
      this.initial = 'Set Value'
    },
    handleInput(val) {
      this.inputValue = val
    },
    validateTodo() {
      this.form.server = this.form.todo
      this.form.todo = ''
    }
  },
  created() {
    this.yolo = 0
    setInterval(() => {
      this.yolo++
    }, 100)
  }
}
</script>

<style lang="scss" scoped>
.create {
  padding: 5px;
  background: gray;
  & > div {
    margin: auto;
  }
}

.list {
  padding: 16px;
  .md-card {
    width: 320px;
    margin: 4px;
  }
}
</style>
```

**Comment by rayfoss** [comment]
cff0929 had a problem clearing fields... setting them to an empty string. Again, what's the justification for making an @input event fire when the model is changed upstream? That's not how input currently works. Regardless, the concession is included.

everything


```
<template>
  <div>
    <md-toolbar class="md-primary">
        <h1 class="md-title">vue-material T0-D0's</h1>
    </md-toolbar>

     <form class="create md-layout" @submit.prevent="validateTodo">
      <md-card class="md-layout-item md-size-50 md-small-size-100">
        <md-card-header>
          <div class="md-title">To-Do next: {{form.server}}</div>
        </md-card-header>

        <md-card-content>
          <div class="md-layout md-gutter">
            <div class="md-layout-item md-small-size-100">
              <md-field>
                <label for="todo">md-input</label>
                <md-input id="todo" v-model="form.todo" />
              </md-field>
              <div class="md-field md-theme-default">
                <input id="todo" v-model="form.todo" placeholder="html input"/>
              </div>
            </div>
          </div>
        </md-card-content>

        <md-card-actions>
          <md-button type="submit" class="md-primary">Create Todo</md-button>
        </md-card-actions>
      </md-card>
    </form>

    <hr>

    <h1>MdInput: {{executions}}</h1>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="initial" @input="handleInput"></md-input>
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div>
  </div>
</template>

<script>
import '../../dist/vue-material.min.css'
import '../../dist/theme/default.css'

const start = Date.now()

export default {
  name: 'TextFields',
  data() {
    return {
      executions: 0,
      initial: '',
      inputValue: '',
      form: {
        todo: '',
        server: ''
      }
    }
  },
  methods: {
    setValue() {
      console.log(`setValue set ${Date.now() - start}`)
      this.executions++
      this.initial = 'Set Value'
    },
    handleInput(val) {
      console.log(`handleInput set ${val} ${Date.now() - start}`)
      this.executions++
      this.inputValue = val
    },
    validateTodo() {
      this.form.server = this.form.todo
      this.form.todo = ''
    }
  }
}
</script>

<style lang="scss" scoped>
.create {
  padding: 5px;
  background: gray;
  & > div {
    margin: auto;
  }
}

.list {
  padding: 16px;
  .md-card {
    width: 320px;
    margin: 4px;
  }
}
</style>

```

**Comment by rayfoss** [comment]
Firefox Android:
screen shot 2018-07-26 at 9 14 39 am
Chrome Android
screen shot 2018-07-26 at 9 13 01 am
Safari
screen shot 2018-07-26 at 9 11 58 am
Chrome
screen shot 2018-07-26 at 9 11 19 am

**Comment by rayfoss** [comment]
Looks good to go:

iPhone:
screen shot 2018-07-26 at 9 30 33 am
Internet exploder
screen shot 2018-07-26 at 9 27 02 am
Micro soft edge
screen shot 2018-07-26 at 9 24 57 am

**Comment by VdustR** [comment]
Your last change will make `<md-input>` without `v-model` dead(can't be modified).
Making data binding non-required is the feature for field components.

Try:
```html
    <md-field>
      <label>Initial Value</label>
      <md-input></md-input>
    </md-field>
```

**Comment by rayfoss** [comment]
@VdustR I've submitted test cases for all the cases we've ran into here.

@marcosmoura I know this is a bit different, but loading mini apps (fixtures) makes testing events and vue in general a heck of a lot easier. I'm for anything that leads to more and better tests in this project. Next, use this technique to add tests to MdTable? it needs them badly

**Comment by rayfoss** [comment]
@Samuell1 @VdustR Try to break it :)

MdSelect needs a major overhaul. MdTextArea needs a couple changes. But given the scope of this issue, I don't want to open that can of worms until we need to. Specially as there are No tests at all for those two components.

If you would prefer I do some light refactoring across about 25 lines of code in MdTextArea and MdSelect, you could say so... but I really feel that's better left for another bug fix pull request or a dedicated refactoring effort.

Aside from the tests, here's my latest pinata 

docs/app/App.vue
```
<template>
  <div>
    <md-toolbar class="md-primary">
        <h1 class="md-title">vue-material T0-D0's</h1>
    </md-toolbar>

     <form class="create md-layout" @submit.prevent="validateTodo">
      <md-card class="md-layout-item md-size-50 md-small-size-100">
        <md-card-header>
          <div class="md-title">To-Do next: {{form.server}}</div>
        </md-card-header>

        <md-card-content>
          <div class="md-layout md-gutter">
            <div class="md-layout-item md-small-size-100">
              <md-field>
                <label for="todo">md-input</label>
                <md-input id="todo" v-model="form.todo" />
              </md-field>
              <div class="md-field md-theme-default">
                <input id="todo" v-model="form.todo" placeholder="html input"/>
              </div>
            </div>
          </div>
        </md-card-content>

        <md-card-actions>
          <md-button type="submit" class="md-primary">Create Todo</md-button>
        </md-card-actions>
      </md-card>
    </form>

    <hr>

    <h1>MdInput: {{executions}}</h1>
    <md-field>
      <label>Initial Value</label>
      <md-input v-model="initial" @input="handleInput"></md-input>
    </md-field>
    <md-button @click="setValue">click</md-button>
    <div>result: {{inputValue}}</div>

    <hr>

    <h1>Non binded empty</h1>
    <md-field>
      <label>Initial Value</label>
      <md-input></md-input>
    </md-field>

    <hr>

    <h1>MdFile Initial: {{fileInitial}}</h1>
    <md-field>
      <label>Initial value</label>
      <md-file v-model="fileInitial" />
    </md-field>

    <hr>

    <h1>MdTextarea Initial: {{textarea}}</h1>
    <md-field>
      <md-textarea v-model="textarea"></md-textarea>
    </md-field>

    <hr>

    <h1>Select Multiple: {{ selectedMovies }}</h1>
    <md-field>
      <label for="movies">Movies</label>
      <md-select v-model="selectedMovies" name="movies" id="movies" multiple>
        <md-option value="fight-club">Fight Club</md-option>
        <md-option value="godfather">Godfather</md-option>
        <md-option value="godfather-ii">Godfather II</md-option>
        <md-option value="godfather-iii">Godfather III</md-option>
        <md-option value="godfellas">Godfellas</md-option>
        <md-option value="pulp-fiction">Pulp Fiction</md-option>
        <md-option value="scarface">Scarface</md-option>
      </md-select>
    </md-field>
  </div>
</template>

<script>
import '../../dist/vue-material.min.css'
import '../../dist/theme/default.css'

const start = Date.now()

export default {
  name: 'TextFields',
  data() {
    return {
      executions: 0,
      initial: '',
      inputValue: '',
      form: {
        todo: '',
        server: ''
      },
      fileInitial: 'vue-material-is-awesome.jpg',
      textarea: '',
      selectedMovies: []
    }
  },
  methods: {
    setValue() {
      console.log(`setValue set ${Date.now() - start}`)
      this.executions++
      this.initial = 'Set Value'
    },
    handleInput(val) {
      console.log(`handleInput set ${val} ${Date.now() - start}`)
      this.executions++
      this.inputValue = val
    },
    validateTodo() {
      this.form.server = this.form.todo
      this.form.todo = ''
    }
  }
}
</script>

<style lang="scss" scoped>
.create {
  padding: 5px;
  background: gray;
  & > div {
    margin: auto;
  }
}

.list {
  padding: 16px;
  .md-card {
    width: 320px;
    margin: 4px;
  }
}
</style>
```

**Comment by rayfoss** [comment]
@Samuell1 @VdustR Would you guys prefer I remove the all tests? This is 15+/15- change but the tests make it seem bigger then it is.

**Comment by Samuell1** [comment]
@rayfoss Yes i think it will be better, to separate to 2 PRs.

**Comment by rayfoss** [comment]
I'll add the tests when the code in the dev branch actually passes them... meanwhile, working on MdField and company is extremely difficult without them, please merge

**Comment by rayfoss** [comment]
Need anything from me? I'd really like to get this fixed so I can actually use it more extensively. This is an extremely common and dangerous bug

**Pull request closed without merging**